### DISCUSSION TITLE
Spock 2.0 JDK compatibility

### DISCUSSION THREAD

**Discussion by lostiniceland** [spockframework/spock#1348] [Q&A] [answered]
I am currently testing Spock 2 and it seems Spock is not compatible with a JDK newer than 11 (currently testing with 16).

My sut is compiled with Java 16 and when compiling the test sources I get "Unsupported class file major version 60" for the sut class.

Since Groovy 3 is compatible with Java 16 I expected Spock to be as well.
```
* What went wrong:
Execution failed for task ':backend-quarkus:compileTestGroovy'.
> BUG! exception in phase 'semantic analysis' in source unit '/xxx/JpaMoneyConverterTest.groovy' Unsupported class file major version 60
```

Am I just missing something or is this a know issue?
Thanks

**Comment by szpak** [comment]
Thanks for the report. The Spock 2.0 release has been automatically tested with Java 8, 11 and 15. JDK 16 is problematic due to required Gradle upgrade to 7.x (which could shake the transitive dependencies) or a need to adjust Gradle 6.x to use a separate JDK 16 just to compile and execute tests.

That being said, there could be a problem with JDK 16, however, as you also mentioned Groovy 3.0.8 should be JDK 16 compatible, the project also uses asm 9.1 (JDK 17 compatible), so could you check that:
 - there are no other Groovy versions on you classpath?
 - it works fine (or not) with JDK 15?

**Comment by Vampire** [comment]
I just gave it a try and here it worked fine.
Could you at least provide the full stacktrace, so that we can see where exactly the exception is originating, or even better an MCVE?

**Comment by leonard84** [comment]
The `Unsupported class file major version XX` exceptions, normally happen if you are trying to consume classes that were compiled with a higher JDK version, than the one executing the task. For example, you compiled `main` with 16 but try to compile `test` with 11. However, as @Vampire said without the full Stacktrace or even better an MCVE we can only guess.

**Comment by lostiniceland** [comment]
Thanks for the replies. I was hesistent to create an issue since I expected something might be off.
After a quick check ASM is version 9.1 but for some reason Gradle is downgrading my transitive Groovy dependencies to 3.0.5.

I will do more checks and if it doestn resolve, I will come back with the required info.

**Comment by lostiniceland** [comment, marked as answer]
Issue seems solved. My problem was that the Quarkus-BOM was enforced by using _enforcedPlatform_ and this always downgraded Groovy to 3.0.5. Replacing _enforcedPlatform_ with _platform_ gave me the option to upgrade to 3.0.8
Now I am hitting another issue with Groovy (not Spock related) because my Java sut is compiled with --enable-preview and this seems not be honored by the Groovy compiler.

Thanks for clarifying that Spock supports newer versions, which lead me on the right track

**Reply by szpak** [reply]
Thanks for getting back with your findings.
